### PR TITLE
[W.I.P] [V1.3] Add the calibration tool doc.

### PR DESCRIPTION
**Pull request by guomingz**
Add the calibration tool doc for PaddlePaddle 1.3 release.

### CODE REVIEW

**Comment by guomingz** [comment]
Hi @luotao1 , I've uploaded the doc per  your [suggestion](https://github.com/PaddlePaddle/Paddle/issues/15063). I also added one [W.I.P] tag which means we'll have an internal review firstly.  I'll notify u once it done.

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
+The 8-bits (**INT8**) inference is also known as the Low Precision Inference which could speed up the inference with the lower accuracy loss. It has higher throughput and lower memory requirements compared to FP32. As the PaddlePaddle enables the INT8 inference supporting from the 1.3 release, we release a accuracy tool interface([Calibrator Class](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/contrib/int8_inference/utility.py#L21)) at the same time. This offline tool will generate the initial quantization parameters firstly and save those scaling factors into the model definition file.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:1]: This tool only support resnet50 in this PR

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+## 0.Prerequisite
+This tool is dedicated to run on Intel SkyLake platform which supports the 8-Bit Inference.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:4]: > Intel SkyLake platform

这段描述是否正确。SkyLake platform往后的平台也不支持？

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+## 1.Usage
+1. Set the env with `export FLAGS_use_mkldnn=True` and build the PaddlePaddle.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:7]: 并不是Set the env with `export FLAGS_use_mkldnn=True`后编译出的paddle。 `export FLAGS_use_mkldnn=True`是使用时的命令。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+2. Generate the model definition file with quantized parameters firstly. We provide a template calibration.py to sample and save the model definition for       int8 inference. Below is the full code snippet. You need to save this file to **/path/to/models/fluid/PaddleCV/image_classification**.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:8]: 为什么需要把文件拷贝到models repo目录下。使用这个功能，和models repo无关。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    ```python
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:9]: 请不要大段贴源码，这些源码不可维护，没有验证。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    # yapf: disable
+    add_arg('batch_size',       int,  32,                 "Minibatch size.")
+    add_arg('use_gpu',          bool, True,                "Whether to use GPU or not.")
+    add_arg('class_dim',        int,  1000,                "Class number.")
+    add_arg('image_shape',      str,  "3,224,224",         "Input image size")
+    add_arg('out',              str,  "calibration_out",   "Output INT8 model")
+    add_arg('with_mem_opt',     bool, True,                "Whether to use memory optimization or not.")
+    add_arg('use_train_data',   bool, False,               "Whether to use train data for sampling or not.")
+    add_arg('pretrained_model', str,  None,                "Whether to use pretrained model.")
+    add_arg('model',            str, "SE_ResNeXt50_32x4d", "Set the network to use.")
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:37]: SE_ResNeXt50_32x4d，目前不支持这个模型，请去掉。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        if not os.path.exists(tmp_scale_folder):
+            os.system("mkdir {}".format(tmp_scale_folder))
+
+        image = fluid.layers.data(name='image', shape=image_shape, dtype='float32')
+        label = fluid.layers.data(name='label', shape=[1], dtype='int64')
+
+        # model definition
+        model = models.__dict__[model_name]()
+
+        if model_name is "GoogleNet":
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:75]: 请先去掉googlenet，验证resnet的正确性

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    import argparse
+    import functools
+    from utility import add_arguments, print_arguments
+    import paddle.fluid.core as core
+    import os
+    sys.path.append('/Path/To/python/paddle/fluid/contrib/int8_inference')
+    import int8_inference.utility as ut
+    parser = argparse.ArgumentParser(description=__doc__)
+    add_arg = functools.partial(add_arguments, argparser=parser)
+    # yapf: disable
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:28]: 请评估下面args是否都必须，比如use_gpu, with_mem_opt等。请尽量提供给用户最少选择。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+    def main():
+        args = parser.parse_args()
+        print_arguments(args)
+        eval(args)
+
+    if __name__ == '__main__':
+        main()
+    ```
+    The typical command of this tool is `python calibration.py --model=ResNet50  --batch_size=50  --image_shape=3,224,224  --with_mem_opt=True  --use_gpu=False  --pretrained_model=weights/resnet50 --debug=False`. When this command executed, it will generate the model definition file which contains scaling factor and save it into the specified output folder.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:166]: 对用户来说，以下默认命令即可。
`python calibration.py --model=ResNet50  --batch_size=50  --image_shape=3,224,224 --pretrained_model=weights/resnet50`

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+            else:
+                with profiler.profiler("CPU", sorted_key='total') as cpuprof:
+                    eval(args)
+        else:
+            eval(args)
+
+    if __name__ == '__main__':
+        main()
+   ```
+    The typical command of this tool is ` python eval_tp_with_model.py  --batch_size=50  --image_shape=3,224,224  --with_mem_opt=True  --use_gpu=False --pretrained_model=calibration_out --iterations 1000`. When this command exectued, it will display the accuracy and throughput on the screen.
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:320]: 两段源代码考虑和单测复用。我理解，应该`test_calibration.py`中，调用`calibration.py`和`eval_tp_with_model.py`来验证。

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+            throughput, throughput_skip, latency, latency_skip))
+        sys.stdout.flush()
+
+    def main():
+        args = parser.parse_args()
+        print_arguments(args)
+        if args.profiler:
+            if args.use_gpu:
+                with profiler.cuda_profiler("cuda_profiler.txt", 'csv') as nvprof:
+                    eval(args)
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:310]: 请重新review下代码，删去无用代码，gpu的代码为什么会在这里出现！

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+        skip_batch_num = args.skip_batch_num
+        if skip_batch_num >= args.iterations:
+        print("Please ensure the skip_batch_num less than iterations.")
+        sys.exit(0)
+        image_shape = [int(m) for m in args.image_shape.split(",")]
+
+        if with_memory_optimization:
+            fluid.memory_optimize(fluid.default_main_program())
+
+        place = fluid.CUDAPlace(0) if args.use_gpu else fluid.CPUPlace()
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:237]: GPU的代码为什么会在这里出现！

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    add_arg('use_gpu',          bool, True,                "Whether to use GPU or not.")
+    add_arg('class_dim',        int,  1000,                "Class number.")
+    add_arg('image_shape',      str,  "3,224,224",         "Input image size")
+    add_arg('with_mem_opt',     bool, True,                "Whether to use memory optimization or not.")
+    add_arg('pretrained_model', str,  None,                "Whether to use pretrained model.")
+    add_arg('skip_batch_num',   int,  5,                    "Skip batch num.")
+    add_arg('use_transpiler',   bool, True,                 'Whether to use transpiler.')
+    add_arg('use_fake_data',    bool, False,                'If set, use fake data instead of real data.')
+    add_arg('iterations',	    int,  100,               	'Fake data iterations')
+    add_arg('profiler',	    bool, False,                'If true, do profiling.')
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:206]: 这里需要做profiler么

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+            eval(args)
+
+    if __name__ == '__main__':
+        main()
+   ```
+    The typical command of this tool is ` python eval_tp_with_model.py  --batch_size=50  --image_shape=3,224,224  --with_mem_opt=True  --use_gpu=False --pretrained_model=calibration_out --iterations 1000`. When this command exectued, it will display the accuracy and throughput on the screen.
+
+    Here is the parameter definitions:
+    * --pretrained_model: the folder that contains pretrained INT8 weights and model;
+    * --iterations: equal to test iteration definition, usually its value equal to epoch/batch size
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:324]: 缺结果，和性能数据（速度和精度）

**Review by luotao1 — COMMENTED** [review]
请对README重写（以resnet50为例）：
- README里不应该贴大段的Python源代码，这些代码无法维护和验证正确性。`calibration.py`和`eval_tp_with_model.py`提供单独的python文件，并在CI中验证正确性。
- 需要提供一键运行脚本，包括从下载数据开始的所有流程。方便QA验证以及用户使用
- 缺少结果和性能数据。

File: python/paddle/fluid/contrib/int8_inference/README.md (hunk 1 above)
Comment (guomingz) [reply python/paddle/fluid/contrib/int8_inference/README.md:1]: How about mobilenet v1?

File: python/paddle/fluid/contrib/int8_inference/README.md (hunk 2 above)
Comment (guomingz) [reply python/paddle/fluid/contrib/int8_inference/README.md:4]: > > Intel SkyLake platform
> 
> 这段描述是否正确。SkyLake platform往后的平台也不支持？

sorry i've added the [W.I.P] tag which means we'll start the internal review firstly, then we'll notify baidu to continue the review.

**Comment by guomingz** [comment]
Close this pr as it's not good enough.

**Review by guomingz — COMMENTED** [review]
1. 

> 请对README重写（以resnet50为例）：
> 
> * README里不应该贴大段的Python源代码，这些代码无法维护和验证正确性。`calibration.py`和`eval_tp_with_model.py`提供单独的python文件，并在CI中验证正确性。

这2个代码是例子，主要是demo怎么用calibrator 这个类。不能贴代码的话，上传到paddle里？
如果可以什么路径
> * 需要提供一键运行脚本，包括从下载数据开始的所有流程。方便QA验证以及用户使用
test_calibration.py 做了这件事
> * 缺少结果和性能数据。
会补充在新的pr里

**Comment by guomingz** [comment]
>应该test_calibration.py中，调用calibration.py和eval_tp_with_model.py来验证。
calibration.py 和eval_tp_with_model.py是两个实例文件 和test_calibration。py 没关系！

File: python/paddle/fluid/contrib/int8_inference/README.md [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+            break
+
+        calibrator.generate_quantized_data(sampling_data)
+
+        fluid.io.save_inference_model(int8_model, feeded_var_names, [
+            calibrator.sampling_program.current_block().var(i) for i in fetch_list
+        ], exe, calibrator.sampling_program)
+        print(
+            "Calibration is done and the corresponding files were generated at {}".
+            format(os.path.abspath(args.out)))
```
Comment (luotao1) [inline python/paddle/fluid/contrib/int8_inference/README.md:156]: 121-156行，和 @hshen14 沟通后，会简化为3-4个API，示例如下（请@hshen14 和 @guomingz 确定后更新示例，并调整单测 ）：
```
calibrator = utility.Calibrator(
            program=test_program,
            pretrained_model=pretrained_model,
            iterations=args.iterations,
            algo=args.algo)
for batch_id, data in enumerate(sampling_reader()):
     exe.run(calibrator.Program(),
                   fetch_list=fetch_list,
                   feed=feeder.feed(data))
save_int8_model(int8_model, feeded_var_names, fetch_list, exe, calibrator.Program())
```

Comment (hshen14) [reply python/paddle/fluid/contrib/int8_inference/README.md:156]: @luotao1 收到，谢谢。

Comment (luotao1) [reply python/paddle/fluid/contrib/int8_inference/README.md:156]: @guomingz @hshen14 封装接口可以参考：https://github.com/PaddlePaddle/models/blob/develop/fluid/PaddleCV/object_detection/README_quant.md
https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/contrib/quantize/quantize_transpiler.py
